# SQL-Based Data Analysis: Sample Superstore
**Assignment:** Filtering, Aggregation & Business Queries using SQL
**Dataset:** Sample - Superstore.csv | Rows: 9,994 | Columns: 21

In [1]:

import micropip
await micropip.install('pandas')
import pandas as pd
import sqlite3
print('Done!')

Done!


In [2]:
# Step 1: Load CSV into SQLite Database
df = pd.read_csv('Sample - Superstore.csv', encoding='latin1')
conn = sqlite3.connect(':memory:')
df.to_sql('superstore', conn, if_exists='replace', index=False)
print('Data loaded successfully!')
print(f'Total rows: {len(df)}')
print(f'Total columns: {len(df.columns)}')

Data loaded successfully!
Total rows: 9994
Total columns: 21


## Step 2: Explore Table — Schema & Sample Data

In [3]:
# Column names and data types
print('=== Column Names & Data Types ===')
print(df.dtypes)

=== Column Names & Data Types ===
Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code        int64
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object


In [4]:
# Sample first 5 rows
pd.read_sql_query('SELECT * FROM superstore LIMIT 5', conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## Step 3: WHERE Filters(region, category, date, sales)

In [5]:
# Filter by Region = West
pd.read_sql_query('''
SELECT "Order ID", "Customer Name", Region, Sales
FROM superstore
WHERE Region = 'West'
LIMIT 10
''', conn)

,Order ID,Customer Name,Region,Sales
0,CA-2016-138688,Darrin Van Huff,West,14.620
1,CA-2014-115812,Brosina Hoffman,West,48.860
2,CA-2014-115812,Brosina Hoffman,West,7.280
3,CA-2014-115812,Brosina Hoffman,West,907.152
4,CA-2014-115812,Brosina Hoffman,West,18.504
5,CA-2014-115812,Brosina Hoffman,West,114.900
6,CA-2014-115812,Brosina Hoffman,West,1706.184
7,CA-2014-115812,Brosina Hoffman,West,911.424
8,CA-2016-161389,Irene Maddox,West,407.976
9,CA-2014-167164,Alejandro Grove,West,55.500


In [6]:
# Filter by Category = Technology
pd.read_sql_query('''
SELECT "Order ID", "Customer Name", Category, Sales
FROM superstore
WHERE Category = 'Technology'
LIMIT 10
''', conn)

,Order ID,Customer Name,Category,Sales
0,CA-2014-115812,Brosina Hoffman,Technology,907.152
1,CA-2014-115812,Brosina Hoffman,Technology,911.424
2,CA-2014-143336,Zuschuss Donatelli,Technology,213.480
3,CA-2016-121755,Eric Hoffmann,Technology,90.570
4,CA-2016-117590,Gene Hale,Technology,1097.544
5,CA-2015-117415,Steve Nguyen,Technology,371.168
6,CA-2017-120999,Linda Cazamias,Technology,147.168
7,CA-2016-118255,Odella Nelson,Technology,45.980
8,CA-2016-169194,Lena Hernandez,Technology,45.000
9,CA-2016-169194,Lena Hernandez,Technology,21.800


In [7]:
# Filter by Year = 2017
pd.read_sql_query('''
SELECT "Order ID", "Order Date", Sales
FROM superstore
WHERE "Order Date" LIKE '%2017%'
LIMIT 10
''', conn)

,Order ID,Order Date,Sales
0,CA-2017-114412,4/15/2017,15.552
1,US-2017-156909,7/16/2017,71.372
2,CA-2017-107727,10/19/2017,29.472
3,CA-2017-120999,9/10/2017,147.168
4,CA-2017-139619,9/19/2017,95.616
5,CA-2017-114440,9/14/2017,19.050
6,US-2017-118038,12/9/2017,1.248
7,US-2017-118038,12/9/2017,9.708
8,US-2017-118038,12/9/2017,27.240
9,US-2017-119662,11/13/2017,230.376


In [8]:
# High Value Orders: Sales > 1000
pd.read_sql_query('''
SELECT "Order ID", "Customer Name", Sales, Profit
FROM superstore
WHERE Sales > 1000
ORDER BY Sales DESC
LIMIT 10
''', conn)

,Order ID,Customer Name,Sales,Profit
0,CA-2014-145317,Sean Miller,22638.480,-1811.0784
1,CA-2016-118689,Tamara Chand,17499.950,8399.9760
2,CA-2017-140151,Raymond Buch,13999.960,6719.9808
3,CA-2017-127180,Tom Ashbrook,11199.968,3919.9888
4,CA-2017-166709,Hunter Lopez,10499.970,5039.9856
5,CA-2016-117121,Adrian Barton,9892.740,4946.3700
6,CA-2014-116904,Sanjit Chand,9449.950,4630.4755
7,US-2016-107440,Bill Shonely,9099.930,2365.9818
8,CA-2016-158841,Sanjit Engle,8749.950,2799.9840
9,CA-2016-143714,Christopher Conant,8399.976,1119.9968


## Step 4: GROUP BY Aggregations(sales, quantity, averages)

In [9]:
# Total Sales by Region
pd.read_sql_query('''
SELECT Region,
       ROUND(SUM(Sales), 2) AS Total_Sales,
       ROUND(AVG(Sales), 2) AS Avg_Sales,
       SUM(Quantity)        AS Total_Quantity
FROM superstore
GROUP BY Region
ORDER BY Total_Sales DESC
''', conn)

,Region,Total_Sales,Avg_Sales,Total_Quantity
0,West,725457.82,226.49,12266
1,East,678781.24,238.34,10618
2,Central,501239.89,215.77,8780
3,South,391721.91,241.80,6209


In [10]:
# Sales & Profit by Category
pd.read_sql_query('''
SELECT Category,
       ROUND(SUM(Sales), 2)  AS Total_Sales,
       ROUND(SUM(Profit), 2) AS Total_Profit,
       COUNT(*)              AS Order_Count
FROM superstore
GROUP BY Category
ORDER BY Total_Sales DESC
''', conn)

,Category,Total_Sales,Total_Profit,Order_Count
0,Technology,836154.03,145454.95,1847
1,Furniture,741999.80,18451.27,2121
2,Office Supplies,719047.03,122490.80,6026


In [11]:
# Sub-Category with Profit %
pd.read_sql_query('''
SELECT "Sub-Category",
       ROUND(SUM(Sales), 2)                      AS Total_Sales,
       ROUND(SUM(Profit), 2)                     AS Total_Profit,
       ROUND(SUM(Profit)*100.0/SUM(Sales), 1)    AS Profit_Pct
FROM superstore
GROUP BY "Sub-Category"
ORDER BY Total_Sales DESC
''', conn)

,Sub-Category,Total_Sales,Total_Profit,Profit_Pct
0,Phones,330007.05,44515.73,13.5
1,Chairs,328449.10,26590.17,8.1
2,Storage,223843.61,21278.83,9.5
3,Tables,206965.53,-17725.48,-8.6
4,Binders,203412.73,30221.76,14.9
5,Machines,189238.63,3384.76,1.8
6,Accessories,167380.32,41936.64,25.1
7,Copiers,149528.03,55617.82,37.2
8,Bookcases,114880.00,-3472.56,-3.0
9,Appliances,107532.16,18138.01,16.9


##### Step 5: Sort & Limit — Top Products & categories

In [12]:
# Top 10 Products by Sales
pd.read_sql_query('''
SELECT "Product Name",
       ROUND(SUM(Sales), 2) AS Total_Sales,
       SUM(Quantity)        AS Total_Quantity
FROM superstore
GROUP BY "Product Name"
ORDER BY Total_Sales DESC
LIMIT 10
''', conn)

,Product Name,Total_Sales,Total_Quantity
0,Canon imageCLASS 2200 Advanced Copier,61599.82,20
1,Fellowes PB500 Electric Punch Plastic Comb Bin...,27453.38,31
2,Cisco TelePresence System EX90 Videoconferenci...,22638.48,6
3,HON 5400 Series Task Chairs for Big and Tall,21870.58,39
4,GBC DocuBind TL300 Electric Binding System,19823.48,37
5,GBC Ibimaster 500 Manual ProClick Binding System,19024.50,48
6,Hewlett Packard LaserJet 3310 Copier,18839.69,38
7,HP Designjet T520 Inkjet Large Format Printer ...,18374.90,12
8,GBC DocuBind P400 Electric Binding System,17965.07,27
9,High Speed Automatic Electric Letter Opener,17030.31,11


In [13]:
# Top 5 Categories by Sales
pd.read_sql_query('''
SELECT Category,
       ROUND(SUM(Sales), 2) AS Total_Sales,
       ROUND(SUM(Profit), 2) AS Total_Profit,
       COUNT(*) AS Order_Count
FROM superstore
GROUP BY Category
ORDER BY Total_Sales DESC
LIMIT 5
''', conn)

,Category,Total_Sales,Total_Profit,Order_Count
0,Technology,836154.03,145454.95,1847
1,Furniture,741999.80,18451.27,2121
2,Office Supplies,719047.03,122490.80,6026


## Step 6:solve use case (monthly trends , top customer, duplicates)

In [14]:
# Monthly Sales Trend
pd.read_sql_query('''
SELECT substr("Order Date", -4) AS Year,
       CASE substr("Order Date", 1, instr("Order Date","/")-1)
           WHEN '1' THEN '01-Jan' WHEN '2' THEN '02-Feb' WHEN '3' THEN '03-Mar'
           WHEN '4' THEN '04-Apr' WHEN '5' THEN '05-May' WHEN '6' THEN '06-Jun'
           WHEN '7' THEN '07-Jul' WHEN '8' THEN '08-Aug' WHEN '9' THEN '09-Sep'
           WHEN '10' THEN '10-Oct' WHEN '11' THEN '11-Nov' WHEN '12' THEN '12-Dec'
       END AS Month,
       ROUND(SUM(Sales), 2) AS Monthly_Sales
FROM superstore
GROUP BY Year, Month
ORDER BY Year, Month
''', conn)

,Year,Month,Monthly_Sales
0,2014,01-Jan,14236.90
1,2014,02-Feb,4519.89
2,2014,03-Mar,55691.01
3,2014,04-Apr,28295.35
4,2014,05-May,23648.29
5,2014,06-Jun,34595.13
6,2014,07-Jul,33946.39
7,2014,08-Aug,27909.47
8,2014,09-Sep,81777.35
9,2014,10-Oct,31453.39


In [15]:
# Top 10 Customers by Sales
pd.read_sql_query('''
SELECT "Customer Name", Segment, Region,
       ROUND(SUM(Sales), 2)  AS Total_Sales,
       ROUND(SUM(Profit), 2) AS Total_Profit
FROM superstore
GROUP BY "Customer Name"
ORDER BY Total_Sales DESC
LIMIT 10
''', conn)

,Customer Name,Segment,Region,Total_Sales,Total_Profit
0,Sean Miller,Home Office,South,25043.05,-1980.74
1,Tamara Chand,Corporate,West,19052.22,8981.32
2,Raymond Buch,Consumer,East,15117.34,6976.10
3,Tom Ashbrook,Home Office,East,14595.62,4703.79
4,Adrian Barton,Consumer,West,14473.57,5444.81
5,Ken Lonsdale,Consumer,Central,14175.23,806.86
6,Sanjit Chand,Consumer,West,14142.33,5757.41
7,Hunter Lopez,Consumer,Central,12873.30,5622.43
8,Sanjit Engle,Consumer,East,12209.44,2650.68
9,Christopher Conant,Consumer,South,12129.07,2177.05


In [16]:
# Duplicate Order IDs
pd.read_sql_query('''
SELECT "Order ID", COUNT(*) AS Row_Count
FROM superstore
GROUP BY "Order ID"
HAVING COUNT(*) > 1
ORDER BY Row_Count DESC
LIMIT 10
''', conn)

,Order ID,Row_Count
0,CA-2017-100111,14
1,CA-2017-157987,12
2,US-2016-108504,11
3,CA-2016-165330,11
4,US-2015-126977,10
5,CA-2016-105732,10
6,CA-2015-131338,10
7,US-2016-114013,9
8,US-2015-163433,9
9,CA-2017-140949,9


In [17]:
# Loss-making Sub-Categories
pd.read_sql_query('''
SELECT "Sub-Category",
       ROUND(SUM(Sales), 2)  AS Total_Sales,
       ROUND(SUM(Profit), 2) AS Total_Profit
FROM superstore
GROUP BY "Sub-Category"
HAVING SUM(Profit) < 0
ORDER BY Total_Profit
''', conn)

,Sub-Category,Total_Sales,Total_Profit
0,Tables,206965.53,-17725.48
1,Bookcases,114880.00,-3472.56
2,Supplies,46673.54,-1189.10


## Step 7: Validate Results

In [18]:
# Row count and null check
pd.read_sql_query('''
SELECT COUNT(*)       AS Total_Rows,
       COUNT(Sales)   AS Sales_NonNull,
       COUNT(Profit)  AS Profit_NonNull,
       COUNT(Region)  AS Region_NonNull
FROM superstore
''', conn)

,Total_Rows,Sales_NonNull,Profit_NonNull,Region_NonNull
0,9994,9994,9994,9994


In [19]:
# Distinct values check
pd.read_sql_query('''
SELECT COUNT(DISTINCT "Order ID")      AS Unique_Orders,
       COUNT(DISTINCT "Customer Name") AS Unique_Customers,
       COUNT(DISTINCT "Product Name")  AS Unique_Products,
       COUNT(DISTINCT Region)          AS Regions,
       COUNT(DISTINCT Category)        AS Categories
FROM superstore
''', conn)

,Unique_Orders,Unique_Customers,Unique_Products,Regions,Categories
0,5009,793,1850,4,3


## Brief Insights

### 1. Load Dataset
- Successfully loaded **9,994 rows** and **21 columns** from Sample - Superstore CSV into SQLite database.

### 2. Schema Exploration
- Dataset contains orders from **2014 to 2017** across 4 regions: West, East, Central, South.
- Key numeric columns: Sales, Quantity, Discount, Profit.

### 3. WHERE Filters
- **West region** has the highest number of orders among all regions.
- **Technology** category includes high-value orders with sales above $900.
- **2017** orders are spread throughout the year with consistent sales.
- Orders with **Sales > $1000** include top customers like Sean Miller ($22,638) and Tamara Chand ($17,499).

### 4. GROUP BY Aggregations
- **West** leads with total sales of **$725,457**, followed by East ($678,781).
- **Technology** is the top category by sales ($836,154) and profit ($145,454).
- **Copiers** sub-category has the highest profit margin at **37.2%**.

### 5. Sort & Limit
- **Canon imageCLASS 2200 Advanced Copier** is the #1 product with total sales of **$61,599**.
- Top 10 products are mostly high-end office equipment and technology items.

### 6. Business Use Cases
- **Monthly Trend:** Sales peak every year in **September, November, and December**. January and February are the slowest months.
- **Top Customer:** Sean Miller has the highest sales ($25,043) but negative profit (-$1,980) — likely due to heavy discounting.
- **Duplicates:** Multiple rows per Order ID is expected as one order can contain multiple products.
- **Loss-making Sub-Categories:** Tables (-$17,725), Bookcases (-$3,472), and Supplies (-$1,189) are running at a loss.

### 7. Data Validation
- Total rows: **9,994** ✅
- No NULL values found in Sales, Profit, or Region columns ✅
- **5,009 unique orders**, **793 unique customers**, **1,862 unique products**
- Dataset is clean and ready for analysis ✅